source/opt/ir_builder.h

```diff
@@ -572,6 +596,20 @@ class InstructionBuilder {
     return AddInstruction(std::move(new_inst));
   }
 
+  Instruction* AddDecoration(uint32_t target_id, spv::Decoration d,
+                             const std::vector<uint32_t>& literals) {
+    std::vector<Operand> operands;
+    operands.push_back({SPV_OPERAND_TYPE_ID, {target_id}});
+    operands.push_back({SPV_OPERAND_TYPE_DECORATION, {uint32_t(d)}});
+    for (uint32_t literal : literals) {
+      operands.push_back({SPV_OPERAND_TYPE_LITERAL_INTEGER, {literal}});
+    }
+
+    std::unique_ptr<Instruction> new_inst(
+        new Instruction(GetContext(), spv::Op::OpDecorate, 0, 0, operands));
```

Comment:
Shall the decoration_manager get an update when doing this?

Reply:
The decoration analysis is not preserved.

In fact, the GetPreservedAnalysis method is not implemented, so everything gets thrown away anyway!

I mean for it to preserve def-use and types.  But looks like there is not much more to do to preserve decoration analysis.

Reply:
I've been trying to update the code to keep the def_use_manager and type manager up to date (and setting preserved analyses).  There's a bunch that needs to be fixed/updated.  So I'll be a little while more

Reply:
oh, this is in the ir builder. !!  Yes, you're right.